File: src/drivers/hd44780.c
//-----------------------------------------------------------------------------
/*

HD44780 LCD Controller Driver

https://en.wikipedia.org/wiki/Hitachi_HD44780_LCD_controller

Supports:

1 row x 16 col
1 row x 20 col
1 row x 24 col
1 row x 40 col

2 row x 16 col
2 row x 20 col
2 row x 24 col
2 row x 40 col

4 row x 16 col
4 row x 20 col

*/
//-----------------------------------------------------------------------------

#include <string.h>

#include "soc.h"
#include "hd44780.h"

//-----------------------------------------------------------------------------

#define LCD_FUNCTION_SET    (0x20 | (0 << 4 /*DL*/) | (1 << 3 /*N*/) | (0 << 2 /*F*/))
#define LCD_DISPLAY_ON      (0x08 | (1 << 2 /*D*/) | (0 << 1 /*C*/) | (0 << 0 /*B*/))
#define LCD_DISPLAY_CLEAR   (0x01)
#define LCD_ENTRY_MODE_SET  (0x04 | (1 << 1 /*I/D*/) | (0 << 0 /*S*/))

//#define LCD_DDRAM_ADR(x)    (0x80 | x)
//#define LCD_HOME            (0x02)
//#define LCD_ROW0            LCD_DDRAM_ADR(0)
//#define LCD_ROW1            LCD_DDRAM_ADR(0x40)

//-----------------------------------------------------------------------------

static void hd44780_wr4(struct hd44780_ctrl *ctrl, uint8_t val) {
	ctrl->clk_hi();
	ctrl->wr(val & 15);
	usDelay(5);
	ctrl->clk_lo();
	usDelay(50);
}

static void hd44780_wr(struct hd44780_ctrl *ctrl, uint8_t val) {
	if (ctrl->mode == HD44780_MODE4) {
		// write 2 x 4 bits
		hd44780_wr4(ctrl, val >> 4);
		hd44780_wr4(ctrl, val);
	} else {
		// write 1 x 8 bits
		ctrl->clk_hi();
		ctrl->wr(val);
		usDelay(5);
		ctrl->clk_lo();
		usDelay(50);
	}
}

// write 8 bits to the instruction/command register
static void hd44780_cmd(struct hd44780_ctrl *ctrl, uint8_t cmd) {
	ctrl->rs_lo();
	usDelay(5);
	hd44780_wr(ctrl, cmd);
}

// write 8 bits to the data register
static void hd44780_data(struct hd44780_ctrl *ctrl, uint8_t ch) {
	ctrl->rs_hi();
	usDelay(5);
	hd44780_wr(ctrl, ch);
}

//-----------------------------------------------------------------------------

// set the row/col cursor position
int hd44780_cursor(struct hd44780_ctrl *ctrl, uint8_t row, uint8_t col) {
	if (ctrl == NULL || row >= ctrl->rows || col >= ctrl->cols) {
		return -1;
	}
	switch (row) {
	case 0:
		if (ctrl->rows == 1) {
			if (col < (ctrl->cols >> 1)) {
				hd44780_cmd(ctrl, 0x80 + col);
			} else {
				hd44780_cmd(ctrl, 0xc0 + col - (ctrl->cols >> 1));
			}
		} else {
			hd44780_cmd(ctrl, 0x80 + col);
		}
		break;
	case 1:
		hd44780_cmd(ctrl, 0xc0 + col);
		break;
	case 2:
		hd44780_cmd(ctrl, 0x80 + ctrl->cols + col);
		break;
	case 3:
		hd44780_cmd(ctrl, 0xc0 + ctrl->cols + col);
		break;
	}
	return 0;
}

// clear the display
void hd44780_clr(struct hd44780_ctrl *ctrl) {
	hd44780_cmd(ctrl, LCD_DISPLAY_CLEAR);
	msDelay(2);
}

// clear a display row
void hd44780_clr_row(struct hd44780_ctrl *ctrl, uint8_t row) {
	// start of the row
	if (hd44780_cursor(ctrl, row, 0) < 0) {
		return;
	}
	// write out space characters
	for (uint8_t i = 0; i < ctrl->cols; i++) {
		hd44780_data(ctrl, ' ');
	}
	// back to the start of the row
	hd44780_cursor(ctrl, row, 0);
}

// display a character
void hd44780_putc(struct hd44780_ctrl *ctrl, uint8_t row, uint8_t col, char c) {
	if (hd44780_cursor(ctrl, row, col) < 0) {
		return;
	}
	hd44780_data(ctrl, c);
}

// display a string
void hd44780_puts(struct hd44780_ctrl *ctrl, uint8_t row, uint8_t col, char *s) {
	if (hd44780_cursor(ctrl, row, col) < 0 || s == NULL) {
		return;
	}
	uint8_t i = col;
	while ((i < ctrl->cols) && (*s != 0)) {
		hd44780_data(ctrl, *s++);
		i++;
	}
}

//-----------------------------------------------------------------------------

// 4-bit mode initialisation
static void hd44780_4bit_init(struct hd44780_ctrl *ctrl) {
	msDelay(20);
	hd44780_wr4(ctrl, 3);
	msDelay(10);
	hd44780_wr4(ctrl, 3);
	msDelay(1);
	hd44780_wr4(ctrl, 3);
	hd44780_wr4(ctrl, 2);

	hd44780_cmd(ctrl, LCD_FUNCTION_SET);
	hd44780_cmd(ctrl, LCD_DISPLAY_ON);
	hd44780_cmd(ctrl, LCD_DISPLAY_CLEAR);
	msDelay(2);

	hd44780_cmd(ctrl, LCD_ENTRY_MODE_SET);
}

// 8-bit mode initialisation
static void hd44780_8bit_init(struct hd44780_ctrl *ctrl) {

	msDelay(20);
	hd44780_wr(ctrl, 0x30);
	msDelay(10);
	hd44780_wr(ctrl, 0x30);
	msDelay(1);
	hd44780_wr(ctrl, 0x30);

	// TODO
}

int hd44780_init(struct hd44780_ctrl *ctrl) {
	if (ctrl == NULL) {
		return -1;
	}
	// initialise the display
	if (ctrl->mode == HD44780_MODE4) {
		hd44780_4bit_init(ctrl);
	} else {
		hd44780_8bit_init(ctrl);
	}
	return 0;
}

//-----------------------------------------------------------------------------

File: og/hd44780.py
# driver for a text lcd with a hd44780 chip

# provide board pin numbers as a tuple in order:
# Rs, E, D4, D5, D6, D7 (4 bit mode)
# Rs, E, D0, D1, D2, D3, D4, D5, D6, D7 (8 bit mode)

from machine import Pin
import utime as time

LCD_DISPLAY_ON = 0x08 | (1 << 2) | (0 << 1) | (0 << 0)
LCD_DISPLAY_OFF = 0x08 | (0 << 2) | (0 << 1) | (0 << 0)
LCD_DISPLAY_CLEAR = 0x01
LCD_ENTRY_MODE_SET = 0x04 | (1 << 1) | (0 << 0)


def LCD_FUNCTION_SET(dl, n):
    return 0x20 | ((dl) << 4) | ((n) << 3) | (0 << 2)


class LCD:
    def __init__(self, mode, pinlist, rows, cols):
        assert rows in (1, 2, 4), "bad rows"
        assert cols in (16, 20, 24, 40), "bad cols"
        assert mode in (4, 8), "bad mode"
        assert len(pinlist) == 2 + mode, "bad pinlist"
        self.mode = mode
        self.cols = cols
        self.rows = rows
        # init hardware pins
        self.LCD_RS = Pin(pinlist[0], Pin.OUT)
        self.LCD_E = Pin(pinlist[1], Pin.OUT)
        self.datapins = [Pin(pin_name, Pin.OUT) for pin_name in pinlist[2:]]
        # initialise the hd44780 chip
        if self.mode == 4:
            # 4 bit mode
            time.sleep_ms(16)
            self._wr(3)
            time.sleep_ms(5)
            self._wr(3)
            time.sleep_ms(1)
            self._wr(3)
            self._wr(2)
        else:
            # 8 bit mode
            time.sleep_ms(16)
            self._wr(0x30)
            time.sleep_ms(5)
            self._wr(0x30)
            time.sleep_ms(1)
            self._wr(0x30)
        self.wr_cmd(LCD_FUNCTION_SET((1, 0)[self.mode == 4], (1, 0)[self.rows == 1]))
        self.wr_cmd(LCD_DISPLAY_ON)
        self.clr()
        self.wr_cmd(LCD_ENTRY_MODE_SET)

    def _wr(self, val):  # write 4 or 8 bits
        self.LCD_E.on()
        for pin in self.datapins:
            pin.value(val & 1)
            val >>= 1
        time.sleep_us(5)
        self.LCD_E.off()
        time.sleep_us(50)

    def wr(self, val):  # write 8 bits
        if self.mode == 4:
            # write 2 x 4 bits
            self._wr(val >> 4)
            self._wr(val)
        else:
            # write 1 x 8 bits
            self._wr(val)

    def wr_cmd(self, cmd):  # write command
        self.LCD_RS.off()
        time.sleep_us(5)
        self.wr(cmd)

    def wr_data(self, val):  # write data
        self.LCD_RS.on()
        time.sleep_us(5)
        self.wr(val)

    def set_cursor(self, row, col):  # set the cursor position
        assert row < self.rows and row >= 0, "bad row"
        assert col < self.cols and col >= 0, "bad col"
        if row == 0:
            self.wr_cmd(0x80 + col)
        elif row == 1:
            self.wr_cmd(0xC0 + col)
        elif row == 2:
            self.wr_cmd(0x80 + self.cols + col)
        elif row == 3:
            self.wr_cmd(0xC0 + self.cols + col)

    def symbol(self, id, data): # user defined symbol
        self.wr_cmd(0x40 + ((id & 7) << 3))
        for i in range(8):
            self.wr_data(data[i])

    def clr(self):  # clear the display
        self.wr_cmd(LCD_DISPLAY_CLEAR)
        time.sleep_ms(2)

    def on(self):  # turn the display on
        self.wr_cmd(LCD_DISPLAY_ON)

    def off(self):  # turn the display off
        self.wr_cmd(LCD_DISPLAY_OFF)

    def clr_row(self, row):  # clear a display row
        self.set_cursor(row, 0)
        for i in range(self.cols):
            self.wr_data(" ")
        self.set_cursor(row, 0)

    def putc(self, row, col, c):  # display a character
        self.set_cursor(row, col)
        self.wr_data(ord(c))

    def puts(self, row, col, s):  # display a string
        self.set_cursor(row, col)
        n = min(len(s), self.cols - col)
        for i in range(n):
            self.wr_data(ord(s[i]))

    def test(self):  # show a test pattern
        self.clr()
        k = 0
        for i in range(self.rows):
            for j in range(self.cols):
                self.putc(i, j, chr(ord("A") + (k % 26)))
                k += 1
